### PR TITLE
Change optimizer tests so that they also always test the peephole optimizer.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
This still needs the shift fix.

File: test/libsolidity/SolidityOptimizer.cpp [hunk 1]
```diff
@@ -665,6 +665,42 @@ BOOST_AUTO_TEST_CASE(optimise_constant_to_codecopy)
...
+	char const* sourceCode = R"(
+		contract C
+		{
+			function f(bytes32 x) public returns (byte r)
+			{
+				assembly { r := and(byte(x, 31), 0xff) }
+			}
+		}
+	)";
+	compileBothVersions(sourceCode);
```
Comment (erak) [inline test/libsolidity/SolidityOptimizer.cpp:679]: Do you think it makes sense to think about support for such comparisons in `soltest`? We could introduce an option (configuration of / options for tests need to be designed, though) and do the comparison plus the additional checks in the framework.

**Review by erak — COMMENTED** [review]
Tests are failing.

File: test/libsolidity/SolidityOptimizer.cpp (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/SolidityOptimizer.cpp:679]: The framework tests against the actual value and is run on both optimized and non-optimized code, so it already does that. We cannot compare the result of other calls because we cannot easily determine which other functions and valid input values exist.

File: test/ExecutionFramework.cpp [hunk 2]
```diff
@@ -57,7 +57,7 @@ ExecutionFramework::ExecutionFramework():
 ExecutionFramework::ExecutionFramework(string const& _ipcPath):
 	m_rpc(RPCSession::instance(_ipcPath)),
 	m_evmVersion(dev::test::Options::get().evmVersion()),
-	m_optimize(dev::test::Options::get().optimize),
+	m_optimiserSettings(dev::test::Options::get().optimize ? solidity::OptimiserSettings::minimal() : solidity::OptimiserSettings::enabled()),
```
Comment (erak) [inline test/ExecutionFramework.cpp]: Not part of this PR, but I find `OptimizerSettings::{minimal, enabled, full}` a bit confusing. And that's shown in this line. I think we could streamline it by changing `enabled` to `default`: `OptimizerSettings::{minimal, default, full}`

Comment (chriseth) [reply test/ExecutionFramework.cpp]: Hm, but the default is minimal ;)

Comment (chriseth) [reply test/ExecutionFramework.cpp]: `regular`?

**Comment by chriseth** [comment]
I will rebase to check for the recently merged Prs.

File: test/ExecutionFramework.cpp (hunk 2 above)
Comment (erak) [reply test/ExecutionFramework.cpp]: `OptimizerSettings::{default, regular, full}` maybe?

**Comment by leonardoalt** [comment]
Needs rebase

File: test/ExecutionFramework.cpp (hunk 2 above)
Comment (leonardoalt) [inline test/ExecutionFramework.cpp]:
```suggestion
	m_optimiserSettings(dev::test::Options::get().optimize ? solidity::OptimiserSettings::enabled()) : solidity::OptimiserSettings::minimal(),
```
Shouldn't it be swapped like this?

Comment (chriseth) [reply test/ExecutionFramework.cpp]: -> https://github.com/ethereum/solidity/pull/6334

File: test/libsolidity/SolidityOptimizer.cpp [hunk 3]
```diff
@@ -54,13 +54,13 @@ class OptimizerTestFramework: public SolidityExecutionFramework
 		unsigned const _optimizeRuns = 200
 	)
 	{
-		bool const c_optimize = m_optimize;
-		unsigned const c_optimizeRuns = m_optimizeRuns;
-		m_optimize = _optimize;
-		m_optimizeRuns = _optimizeRuns;
+		OptimiserSettings previousSettings = std::move(m_optimiserSettings);
+		// This uses "none" / "full" while most other test frameworks use
```
Comment (erak) [inline test/libsolidity/SolidityOptimizer.cpp:58]: This comment should be changed with https://github.com/ethereum/solidity/pull/6334.

**Review by erak — APPROVED** [review]
